controller_connections.go

```diff
@@ -90,19 +92,135 @@ func (c *controller) handleIncoming(con net.Conn) {
+// case this function returns an error AND a list of alternate endpoints
+func (c *controller) handshake(ip string, con net.Conn, incoming bool) (*Peer, []Endpoint, error) {
+	tmplogger := c.logger.WithField("addr", ip).WithField("incoming", incoming)
+	timeout := time.Now().Add(c.net.conf.HandshakeTimeout)
+
+	nonce := make([]byte, 8) // loopback detection
+	binary.LittleEndian.PutUint64(nonce, c.net.instanceID)
+
+	// upgrade connection to a metrics connection
+	metrics := NewMetricsReadWriter(con)
+
+	handshake := newHandshake(c.net.conf, nonce)
+	decoder := gob.NewDecoder(metrics) // pipe gob through the metrics writer
+	encoder := gob.NewEncoder(metrics)
+	con.SetWriteDeadline(timeout)
+	con.SetReadDeadline(timeout)
+
+	failfunc := func(err error) (*Peer, []Endpoint, error) {
+		tmplogger.WithError(err).Debug("Handshake failed")
+		con.Close()
+		return nil, nil, err
+	}
+
+	err := encoder.Encode(handshake)
+	if err != nil {
+		return failfunc(fmt.Errorf("Failed to send handshake"))
+	}
+
+	var reply Handshake
+	err = decoder.Decode(&reply)
+	if err != nil {
+		return failfunc(fmt.Errorf("Failed to read handshake"))
+	}
+
+	// check basic structure
+	if err = reply.Valid(c.net.conf); err != nil {
+		return failfunc(err)
+	}
+
+	endpoint, err := NewEndpoint(ip, reply.Header.PeerPort)
+	if err != nil {
+		return failfunc(fmt.Errorf("failed to create endpoint: %v", err))
+	}
+
+	if c.isBannedEndpoint(endpoint) {
+		return failfunc(fmt.Errorf("Peer %s is banned, disconnecting", endpoint))
+	}
+
+	// loopback detection
+	if bytes.Equal(reply.Payload, nonce) {
+		return failfunc(fmt.Errorf("loopback"))
+	}
 
-	if c.isBannedEndpoint(peer.Endpoint) {
-		c.logger.Debugf("Peer %s is banned, disconnecting", peer.Hash)
-		peer.Stop()
+	prot, err := c.determineProtocol(&reply, con, decoder, encoder)
+	if err != nil {
+		return failfunc(err)
+	}
+
+	// dialed a node that's full
+	if reply.Header.Type == TypeRejectAlternative {
+		con.Close()
+		tmplogger.Debug("con rejected with alternatives")
+
+		share, err := prot.ParsePeerShare(reply.Payload)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to parse alternatives: %s", err.Error())
+		}
+
+		for _, ep := range share {
+			if !ep.Valid() {
+				return nil, nil, fmt.Errorf("peer provided invalid peer share")
+			}
+		}
+		return nil, share, fmt.Errorf("connection rejected")
+	}
+
+	peer := newPeer(c.net, uint32(reply.Header.NodeID), endpoint, con, prot, metrics, incoming)
+
+	c.peerStatus <- peerStatus{peer: peer, online: true}
+
+	// a p2p1 node sends a peer request, so it needs to be processed
+	if reply.Header.Type == TypePeerRequest {
+		req := newParcel(TypePeerRequest, []byte("Peer Request"))
+		req.Address = peer.Hash
+		c.peerData <- peerParcel{peer: peer, parcel: req}
+	}
+
+	return peer, nil, nil
+}
+
+func (c *controller) determineProtocol(hs *Handshake, conn net.Conn, decoder *gob.Decoder, encoder *gob.Encoder) (Protocol, error) {
```

Comment:
nitpicking but I would pass strictly just what this function need, so just the version instead of the whole handshake